### PR TITLE
feat(server): support use other server frameworks, like express

### PR DESCRIPTION
**Pull request by 9aoy**
## Summary

rsbuild supports start with other server frameworks, like [express](https://expressjs.com/) (This API is not stable)

example: /e2e/cases/server/custom-server/scripts/server.mjs
## Checklist

- [ ] Tests updated.
- [ ] Documentation updated.

### CODE REVIEW

File: e2e/cases/server/custom-server/scripts/server.mjs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  const httpServer = app.listen({ host, port }, async () => {
+    await rsbuildServer.afterStart({
+      port,
+      routes: defaultRoutes,
+    });
+  });
+
+  // subscribe the server's http upgrade event to handle WebSocket upgrade
+  httpServer.on('upgrade', upgrade);
```
Comment (chenjiahan) [inline e2e/cases/server/custom-server/scripts/server.mjs:47]: upgrade -> onUpgrade?

File: e2e/cases/server/custom-server/scripts/server.mjs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      plugins: [pluginReact()],
+      server: {
+        htmlFallback: false,
+      },
+    },
+  });
+
+  const app = express();
+
+  const rsbuildServer = await rsbuild.createDevServer();
```
Comment (chenjiahan) [inline e2e/cases/server/custom-server/scripts/server.mjs:19]: createDevServer -> getServerAPIs?

File: e2e/cases/server/custom-server/package.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+  "private": true,
+  "name": "@e2e/custom-server",
+  "version": "1.0.0",
+  "scripts": {
+    "dev": "node ./scripts/index.mjs"
+  },
+  "dependencies": {
+    "express": "^4.18.2"
```
Comment (chenjiahan) [inline e2e/cases/server/custom-server/package.json:9]: express has many dependencies, can we find a lightweight library to test the server ability?

Comment (9aoy) [reply e2e/cases/server/custom-server/package.json:9]: express is already a relatively lightweight dependencies among other common node frameworks (compared to fastify, hapi, restify).

koa is lighter than express, but koa's response is different from node's http response, so it is not supported in rsbuild. 😂

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**